### DISCUSSION TITLE
config migration loses regexManagers when customManagers are present

### DISCUSSION THREAD

**Discussion by candrews** [renovatebot/renovate#32020] [Request Help]
### How are you running Renovate?

A Mend.io-hosted app

### Please tell us more about your question or problem

When the renovate configuration has both `regexManagers` and `customManagers` set, Renovate should migrate both the  `regexManagers` and `customManagers` into `customManagers`. However, it actually just drops the `regexManagers`.

Here's a PR created by Renovate demonstrating this issue: https://github.com/candrews/renovate-config-migration/pull/3

I believe this issue should block enabling `configMigration` by default: https://github.com/renovatebot/renovate/issues/16359

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: Config migration necessary
{
  "oldConfig": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "mergeConfidence:all-badges",
      "config:base"
    ],
    "configMigration": true,
    "regexManagers": [
      {
        "fileMatch": [
          ".*\\.yml$"
        ],
        "matchStrings": [
          "TRIVY_POLICIES_IMAGE *: *\"(?<currentValue>.+?)\""
        ],
        "depNameTemplate": "docker",
        "datasourceTemplate": "docker"
      }
    ],
    "customManagers": [
      {
        "fileMatch": [
          "(^|/)Dockerfile$"
        ],
        "matchStrings": [
          "ARG COSIGN_VERSION=(?<currentValue>.+?)\\n"
        ],
        "depNameTemplate": "sigstore/cosign",
        "datasourceTemplate": "github-releases"
      }
    ]
  }
  "newConfig": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "mergeConfidence:all-badges",
      "config:recommended"
    ],
    "configMigration": true,
    "customManagers": [
      {
        "customType": "regex",
        "fileMatch": [
          "(^|/)Dockerfile$"
        ],
        "matchStrings": [
          "ARG COSIGN_VERSION=(?<currentValue>.+?)\\n"
        ],
        "depNameTemplate": "sigstore/cosign",
        "datasourceTemplate": "github-releases"
      }
    ]
  }
}

DEBUG: Post-massage config
{
  "config": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "mergeConfidence:all-badges",
      "config:recommended"
    ],
    "configMigration": true,
    "customManagers": [
      {
        "customType": "regex",
        "fileMatch": [
          "(^|/)Dockerfile$"
        ],
        "matchStrings": [
          "ARG COSIGN_VERSION=(?<currentValue>.+?)\\n"
        ],
        "depNameTemplate": "sigstore/cosign",
        "datasourceTemplate": "github-releases"
      }
    ]
  }
}
```

</details>

**Comment by rarkins** [comment]
Having both old and new options in the same config should be considered a config error and not a valid use case. I've looked at the migration code we have for such cases and don't see any quick way to handle this, and don't think it's worth our time to figure out how to do do. You are welcome to contribute a PR if you can find an elegant way. The starting point would be https://github.com/renovatebot/renovate/blob/main/lib/config/migrations/base/rename-property-migration.ts

It's also not blocking for enabling configMigration regardless because users are still expected to review the proposed migrations. We can't have a "it must be perfect otherwise we can never enable it" policy.

**Reply by astellingwerf** [reply]
Although I agree with you that it does not technically block #16359 from being enabled by default, the issue the @candrews describes of losing bits of a users custom configuration isn't a good problem to have. Due to the high quality of Renovate, users have come to expect that the PRs generally are in order or fail the build. But neither would be the case here. Enabling this behaviour by default could lower the trust folks have in Renovate.

Even if people would check the PRs changeset, it might not be immediately obvious what's going on. The Config Migration PR does not mention what migrations were applied, nor why. To communicate to the user, might it be useful to commit the output of each migration step separately instead of as one opaque monolithic commit with everything?

>  I've looked at the migration code we have for such cases and don't see any quick way to handle this

Doesn't PackagesMigration (from a logic perspective) do what we're looking for?

**Reply by RahulGautamSingh** [reply]
> the issue the @candrews describes of losing bits of a users custom configuration isn't a good problem to have

Not sure we can call accidental missing of config bits a renovate problem if the user config is invalid to begin with.

We could add a validation step for this instead of migration, because that's what it is. It might also be easier to implement.
